### PR TITLE
testing first service

### PR DESCRIPTION
**Pull request by Andrij-Maksymyshyn**

### CODE REVIEW

File: __tests__/services/usersServices/getSingleUser.test.js [hunk 1]
```diff
@@ -0,0 +1,38 @@
+const User = require("../../../models/user");
+const { getSingleUser } = require("../../../api/users/services");
+
+jest.mock("../../../models/user");
+
+const id = "123";
+
+describe("getSingleUser function", () => {
+  it("should return user, when user exists", async () => {
+    User.findOne.mockResolvedValueOnce(() => ({
```
Comment (chestnyh) [inline __tests__/services/usersServices/getSingleUser.test.js]: Why in this test you use `mockResolvedValueOnce` but in the next one `mockImplementationOnce`.

File: __tests__/services/usersServices/getSingleUser.test.js [hunk 2]
```diff
@@ -0,0 +1,38 @@
...
+
+    await getSingleUser(id);
+
+    expect(User.findOne).toHaveBeenCalledWith({
+      _id: id,
+      isDeleted: false
+    });
+  });
+
+  /////////////////////////////////////////////////////////////////////////////
```
Comment (chestnyh) [inline __tests__/services/usersServices/getSingleUser.test.js]: Strange comment

Comment (chestnyh) [reply __tests__/services/usersServices/getSingleUser.test.js]: Why don't you check what `getSingleUser` return to you

File: __tests__/services/usersServices/getSingleUser.test.js [hunk 3]
```diff
@@ -0,0 +1,38 @@
...
+  it("should return null, when user doesn't exist", async () => {
+    const newId = "12345";
+
+    User.findOne.mockImplementationOnce(() => null);
+
+    await getSingleUser(newId);
```
Comment (chestnyh) [inline __tests__/services/usersServices/getSingleUser.test.js]: Why don't you check what getSingleUser return to you

File: __tests__/services/usersServices/getSingleUser.test.js [hunk 4]
```diff
@@ -0,0 +1,41 @@
...
+    const mockReturn = {
+      _id: id,
+      isDeleted: false
+    };
+
+    User.findOne.mockResolvedValueOnce(() => mockReturn);
+    const result = await getSingleUser(id);
+
+    expect(User.findOne).toHaveBeenCalledWith(mockReturn);
+    expect(result()).toBe(mockReturn);
```
Comment (chestnyh) [inline __tests__/services/usersServices/getSingleUser.test.js]: result() ??? Why `getSingleUser` returns method?

Probably you need to `User.findOne.mockResolvedValueOnce(mockReturn);` in line 20

Comment (chestnyh) [reply __tests__/services/usersServices/getSingleUser.test.js]: The same for the test below

File: api/users/services/addUser.js [hunk 5]
```diff
@@ -2,11 +2,11 @@ const { hashPassword } = require("../../../services");
 const { User } = require("../../../models");
 
 const addUser = async userObject => {
-  const hashedPassword = await hashPassword(userObject.password);
+  const password = await hashPassword(userObject.password);
 
   const newUser = User.create({
```
Comment (chestnyh) [inline api/users/services/addUser.js]: Why this is not with await??

File: __tests__/services/usersServices/addUser.test.js [hunk 6]
```diff
@@ -0,0 +1,32 @@
...
+const { addUser } = require("../../../api/users/services");
+const { hashPassword } = require("../../../services");
+
+jest.mock("../../../models/user");
+
+jest.mock("../../../services", () => ({
+  hashPassword: jest.fn(() => "hash password")
+}));
+
+describe("addUser function", () => {
```
Comment (chestnyh) [inline __tests__/services/usersServices/addUser.test.js:11]: You should write test when mock throws error and how `addUser` behave if we have error there.

File: __tests__/services/usersServices/addUser.test.js [hunk 7]
```diff
@@ -0,0 +1,32 @@
...
+
+  it("should create new user, when we pass the object to create", async () => {
+    const mockedUser = {
+      fullName: "fake_name",
+      email: "fake_email",
+      password: "fake_password",
+      avatarUrl: "fake_url"
+    };
+
+    User.create.mockResolvedValueOnce(() => mockedUser);
```
Comment (chestnyh) [inline __tests__/services/usersServices/addUser.test.js]: The same... Looks like you need to resolve mockUser, not function

File: __tests__/services/usersServices/addUser.test.js [hunk 8]
```diff
@@ -0,0 +1,32 @@
...
+
+    const result = await addUser(mockedUser);
+
+    expect(hashPassword).toHaveBeenCalledWith("fake_password");
```
Comment (chestnyh) [inline __tests__/services/usersServices/addUser.test.js:28]: Why don't you check if `User.create` was called with correct values

File: __tests__/services/usersServices/getSingleUser.test.js (hunk 4 above)
Comment (Andrij-Maksymyshyn) [reply __tests__/services/usersServices/getSingleUser.test.js]: done

File: api/users/services/addUser.js (hunk 5 above)
Comment (Andrij-Maksymyshyn) [reply api/users/services/addUser.js]: done

File: __tests__/services/usersServices/addUser.test.js (hunk 7 above)
Comment (Andrij-Maksymyshyn) [reply __tests__/services/usersServices/addUser.test.js]: done

File: __tests__/services/usersServices/addUser.test.js (hunk 8 above)
Comment (Andrij-Maksymyshyn) [reply __tests__/services/usersServices/addUser.test.js:28]: done

File: __tests__/services/usersServices/addUser.test.js [hunk 9]
```diff
@@ -0,0 +1,59 @@
...
+    expect(User.create).toHaveBeenCalledWith({
+      ...mockedUser,
+      password: "hash password"
+    });
+
+    expect(result).toBe(mockedUser);
+  });
+
+  it("function should throw error, when we didn't pass password", async () => {
+    const addUser = jest.fn(x => x);
```
Comment (aleksei-podvisory) [inline __tests__/services/usersServices/addUser.test.js]: ???? Why you assign jest function to addUser

File: __tests__/services/usersServices/addUser.test.js [hunk 10]
```diff
@@ -0,0 +1,59 @@
...
+
+    addUser.mockResolvedValueOnce(() => {
+      throw new Error("not passed password");
+    });
```
Comment (aleksei-podvisory) [inline __tests__/services/usersServices/addUser.test.js]: Should be mock rejected value

File: __tests__/services/usersServices/addUser.test.js [hunk 11]
```diff
@@ -0,0 +1,59 @@
...
+      avatarUrl: "fake_url"
+    };
+
+    User.create.mockResolvedValueOnce(mockedUser);
+
+    const result = await addUser(mockedUser);
+
+    // expect(User.create).toHaveBeenCalledWith(mockedUser);
+
+    expect(result).toThrow(new Error("not passed password"));
```
Comment (aleksei-podvisory) [inline __tests__/services/usersServices/addUser.test.js]: If you have error in line 53. You can't call line 57. It has no sense.

File: __tests__/services/usersServices/getSingleUser.test.js [hunk 12]
```diff
@@ -0,0 +1,41 @@
...
+      _id: id,
+      isDeleted: false
+    };
+
+    User.findOne.mockResolvedValueOnce(mockReturn);
+    const result = await getSingleUser(id);
+
+    expect(User.findOne).toHaveBeenCalledWith(mockReturn);
+    expect(result).toBe(mockReturn);
+  });
```
Comment (chestnyh) [inline __tests__/services/usersServices/getSingleUser.test.js]:
```
 it("should return user, when user exists", async () => {
    const mockReturn = {
        _id: id,
        name: "Andrii",
        email: "ggg@ggg.ua"
    };

    User.findOne.mockResolvedValueOnce(mockReturn);
    const result = await getSingleUser(id);

    expect(User.findOne).toHaveBeenCalledWith({
      _id: id,
      isDeleted: false
    });
    expect(result).toBe(mockReturn);
  });

File: __tests__/services/usersServices/addUser.test.js (hunk 11 above)
Comment (chestnyh) [reply __tests__/services/usersServices/addUser.test.js]:
```
expect(async() => await addUser(mockedUser)).toThrow(new Error("not passed password"));

File: __tests__/services/usersServices/getAllUsers.test.js [hunk 13]
```diff
@@ -0,0 +1,52 @@
...
+        },
+        {
+          fullName: "user2"
+        },
+        {
+          fullName: "user3"
+        }
+      ],
+      page: 1,
+      perPage: Number(3),
```
Comment (aleksei-podvisory) [inline __tests__/services/usersServices/getAllUsers.test.js]: Why Number(3) here? Why just not 3

File: __tests__/services/usersServices/getAllUsers.test.js [hunk 14]
```diff
@@ -0,0 +1,52 @@
...
+      total: Number(5)
+    };
+
+    User.find.mockResolvedValueOnce(mockedReturn);
+
+    User.count.mockResolvedValueOnce(mockedReturn.total);
+
+    await getAllUsers();
```
Comment (chestnyh) [inline __tests__/services/usersServices/getAllUsers.test.js]: Why don't you check what user returns

File: __tests__/services/usersServices/getAllUsers.test.js [hunk 15]
```diff
@@ -0,0 +1,52 @@
...
+
+    expect(User.find).toHaveBeenCalledWith({ isDeleted: false }, "", {
+      limit: 3,
+      skip: 0
+    });
+
+    expect(User.count).toHaveReturned();
```
Comment (chestnyh) [inline __tests__/services/usersServices/getAllUsers.test.js]: What do you check here?

File: __tests__/services/usersServices/getAllUsers.test.js [hunk 16]
```diff
@@ -0,0 +1,52 @@
...
+const { getAllUsers } = require("../../../api/users/services");
+
+jest.mock("../../../models/user");
+
+describe("getAllUsers function", () => {
+  afterEach(() => {
+    jest.resetAllMocks();
+  });
+
+  it("should return usersData(& usersTotal) with page = 1, perPage = 3, when we didn't pass query", async () => {
```
Comment (chestnyh) [inline __tests__/services/usersServices/getAllUsers.test.js]: You should check `getAllUsers` with params

File: __tests__/helpers/buildSortQuery.test.js [hunk 17]
```diff
@@ -0,0 +1,25 @@
...
+      .mockImplementationOnce(fakeQuery);
+
+    const result = buildSortQuery("asc(title),desc(_id)");
+    console.log("result:", result);
+
+    expect(fakedSortQuery).toHaveBeenCalledWith("asc(title),desc(_id)");
+
+    expect(result).toBe(fakeQuery);
+  });
+});
```
Comment (chestnyh) [inline __tests__/helpers/buildSortQuery.test.js]: const { buildSortQuery } = require("../../helpers/buildSortQuery");

describe("buildSortQuery function", () => {

  it("should return query for sorting, when we passed sorting string", () => {

   const params = "asc(title),desc(_id)";
   const expected = {title: -1, _id: 1}

    const result = buildSortQuery(params);
    expect(JSON.stringify(result)).(JSON.stringify(expected));
  });
});

File: __tests__/helpers/buildFilterQuery.test.js [hunk 18]
```diff
@@ -0,0 +1,43 @@
+const { buildFilterQuery } = require("../../helpers/buildFilterQuery");
+
+describe("buildFilterQuery function", () => {
+  it("should return query for filtering, when we passed query with one param", () => {
+    const fakedQuery = {
```
Comment (chestnyh) [inline __tests__/helpers/buildFilterQuery.test.js]: Not necessary to name it `fakedQuery`,  `query` is fine

File: __tests__/helpers/buildFilterQuery.test.js [hunk 19]
```diff
@@ -0,0 +1,43 @@
...
+      authorId: "123yes"
+    };
+    const expectedFilterQuery = {
```
Comment (chestnyh) [inline __tests__/helpers/buildFilterQuery.test.js]: just `expected` is fine

File: __tests__/helpers/buildSortQuery.test.js [hunk 20]
```diff
@@ -0,0 +1,11 @@
+const { buildSortQuery } = require("../../helpers/buildSortQuery");
+
+describe("buildSortQuery function", () => {
+  it("should return query for sorting, when we passed sorting string", () => {
+    const fakedParams = "asc(a),desc(b)";
+    const expectedSortQuery = { a: "1", b: "-1" };
+
+    const result = buildSortQuery(fakedParams);
+    expect(result).toStrictEqual(expectedSortQuery);
+  });
```
Comment (chestnyh) [inline __tests__/helpers/buildSortQuery.test.js]: Add case when nothing passer as argument

File: __tests__/services/authServices/createOauthPair.test.js [hunk 21]
```diff
@@ -0,0 +1,45 @@
...
+describe("createOauthPair function", () => {
+  afterEach(() => {
+    jest.resetAllMocks();
+  });
+
+  it("should return new tokenPair, when we passed tokenData", async () => {
+    Oauth.create.mockResolvedValueOnce(mokedtokenPair);
+
+    const result = await createOauthPair(tokenDataToCreate);
+    console.log("result:", result);
```
Comment (chestnyh) [inline __tests__/services/authServices/createOauthPair.test.js]: Remove console.log()

File: __tests__/errors/AppiError.test.js [hunk 22]
```diff
@@ -0,0 +1,53 @@
...
+    const obj = new ApiError("Something went wrong", 418);
+    expect(obj.message).toBe("Something went wrong");
+    expect(obj.status).toBe(418);
+  });
+
+  it("should return new instance for BadRequest class with 2 public properies", () => {
+    const obj = new BadRequest("Something went wrong");
+    expect(obj.message).toBe("Something went wrong");
+    expect(obj.status).toBe(400);
+  });
```
Comment (chestnyh) [inline __tests__/errors/AppiError.test.js]: it("should return new instance for BadRequest class with 2 public properies", () => {
    
    const errorMessage = "Something went wrong"

    const obj = new BadRequest(errorMessage);
    expect(obj.message).toBe(errorMessage);
    expect(obj.status).toBe(400);
  });

File: __tests__/helpers/buildFilterQuery.test.js (hunk 18 above)
Comment (Andrij-Maksymyshyn) [reply __tests__/helpers/buildFilterQuery.test.js]: already fixed

File: __tests__/helpers/buildFilterQuery.test.js (hunk 19 above)
Comment (Andrij-Maksymyshyn) [reply __tests__/helpers/buildFilterQuery.test.js]: already fixed

File: __tests__/helpers/buildSortQuery.test.js (hunk 20 above)
Comment (Andrij-Maksymyshyn) [reply __tests__/helpers/buildSortQuery.test.js]: already fixed

File: __tests__/services/authServices/createOauthPair.test.js (hunk 21 above)
Comment (Andrij-Maksymyshyn) [reply __tests__/services/authServices/createOauthPair.test.js]: already fixed

File: __tests__/errors/AppiError.test.js (hunk 22 above)
Comment (Andrij-Maksymyshyn) [reply __tests__/errors/AppiError.test.js]: already fixed

File: __tests__/commonMiddleware.test.js/validateSchema.test.js [hunk 23]
```diff
@@ -0,0 +1,28 @@
+const Joi = require("joi");
+
+const { OBJECT_ID } = require("../../configs/regexp.enum");
+const { validate } = require("../../commonMiddleware");
+const { BAD_REQUEST } = require("../../errors/errorCodes");
+
+describe("validate function", () => {
+  it("should pass next value after validation schema", () => {
+    const req = {
```
Comment (chestnyh) [inline __tests__/commonMiddleware.test.js/validateSchema.test.js:19]: Rename as schema

File: __tests__/commonMiddleware.test.js/validateSchema.test.js [hunk 24]
```diff
@@ -0,0 +1,28 @@
...
+      params: Joi.object().keys({
+        userId: Joi.string().alphanum().regex(OBJECT_ID).required()
+      })
+    };
+
```
Comment (chestnyh) [inline __tests__/commonMiddleware.test.js/validateSchema.test.js:40]: add `const req = ....`

**Review by chestnyh — APPROVED** [review]

**Pull request merged**